WIP - enhance README

fturib opened the pull request:
<!--
Thank you for contributing to CoreDNS!

Please provide the following information to help us make the most of your pull request:
-->

### 1. What does this pull request do?

Add a sub-section in the section "Security" of README, describing the security release process that CoreDNS expect to follow  in order to be compatible with K8s's own security release process

### 2. Which issues (if any) are related?
CoreDNS going GA for k8s.


### 3. Which documentation changes (if any) need to be made?
README.

fturib commented:
@miekg , @johnbelamaric , following recommendation of Tim A. from K8s Team, I propose that we have a description of our security release process.

Right now, I just enhance the README.

Let me know if that is preferable to have a separated document ... and where.

Initial reco from Tim A .is:

> I think the requirements for the CoreDNS security process to be compatible with ours are:
> 
> 1. Disclosures can be made privately
> 2. kubernetes-security@googlegroups.com gets early notification of upcoming security releases
> 3. A compatible public disclosure & release timeline, so we have enough time to coordinate a corresponding Kubernetes release. Specifically:
>   - public releases should correspond with public disclosure (controlled), i.e. within O(hours)
>   - our timelines need to be revisited, but ideally we would have ~1 week to pull the release together. We could probably scramble in 24 hours for a critical emergency.
> 4. Backwards-compatible patch releases are provided for releases going back at least as far as the Kubernetes support timeline (3 releases).

Let me know what you think.

Thanks

miekg commented on README.md:
```diff
@@ -198,3 +198,17 @@ If you find a security vulnerability or any security related issues,
 will publicly thank you for it.
+
+
+### Coordination of security releases with Kubernetes project
```
I have some opinions about this

miekg commented on README.md:
```diff
@@ -198,3 +198,17 @@ If you find a security vulnerability or any security related issues,
+
+### Coordination of security releases with Kubernetes project
+
+As CoreDNS is now a feature part of Kubernetes, we are organizing our security release process 
```
Which is nice and all, but does that more it more special than, say, Cloudflare which also use coredns in some project?

fturib replied:
As you propose below, let's make something more generic with "collaborating projects"

miekg commented on README.md:
```diff
@@ -198,3 +198,17 @@ If you find a security vulnerability or any security related issues,
+As CoreDNS is now a feature part of Kubernetes, we are organizing our security release process 
+in a way that is compatible with [Kubernetes's security release process](https://github.com/kubernetes/sig-elease/blob/master/security-release-process-documentation/security-release-process.md)
+
+1. CoreDNS Team may disclose privately the security issue in order to organize the patch and its release.
```
The paragraph above also talks about the security process, should probably merge these two

miekg commented on README.md:
```diff
@@ -198,3 +198,17 @@ If you find a security vulnerability or any security related issues,
+in a way that is compatible with [Kubernetes's security release process](https://github.com/kubernetes/sig-elease/blob/master/security-release-process-documentation/security-release-process.md)
+
+1. CoreDNS Team may disclose privately the security issue in order to organize the patch and its release.
+1. CoreDNS Team will inform Kubernete's project using kubernetes-security@googlegroups.com as early as possible and coordinate 
```
We should make this generic and create a private security announce where interesting parties should subscribe

fturib replied:
Good point. I agree.

miekg commented on README.md:
```diff
@@ -198,3 +198,17 @@ If you find a security vulnerability or any security related issues,
+can organize its own security release in compatibility with the public disclosure date of the issue. Specifically:
+   * public releases should correspond with public disclosure (controlled), i.e. within O(hours)
+   * ideally CoreDNS warns Kubernetes one week before the security release date. That time could probably be scrambled downto 24 hours for a critical emergency.
+1. CoreDNS Team will provide backwards-compatible patch releases for all of the CoreDNS releases embedded into one of the 3 last releases of Kubernetes.
```
We have no such policy and I think the unofficial one is "upgrade to latest version". I also believe we lack the people power to provide such backwards compatibility.

fturib replied:
So, if we have to provide a security release, it will be on top of the current official release ? 
for instance, right now, it would be on top of 1.1.3.

The collaborating projects would have to release a patch that includes that new release "1.1.3-security-patch-01" ?

fturib commented:
I pushed a second version of the "Security releases coordination"

fturib commented on README.md:
```diff
@@ -199,16 +199,20 @@ please DO NOT file a public issue, instead send your report privately to
+That list aims to coordinate the security release in a controlled manner: 
+ 
+1. When a security issue is raised that triggers a CoreDNS code change and therefore a security release, 
+CoreDNS Team will inform privately each of the Collaborating Projects as early as possible, providing the timeline of CoreDNS release, 
```
Who has a better name for "Collaborating Projects" ? (maybe included Security in that name ..) ?

miekg replied:
I think we should do the opposite and require any interested project to join (a new) mailing list

fturib replied:
I guess, in practical, what would happen is:
1- we receive a vulnerability issue somehow (incl email at security@coredns.io)
2- we assess if that needs a security release of CoreDNS
if YES:
3- we (CoreDNS Team) elaborate a timeline for release (I guess in a way that is compatible with subscribed projects)
4- we then inform the dl-security-release of the issue and our plan to remediate
5- I guess if that needs bargaining for the release timeframe, then we will be contacted by those projects that expect adaptation of the plan.
6- When plan is definitive, resend an email to advertise that plan and ETA release for CoreDNS.

Would that work ?

fturib commented on README.md:
```diff
@@ -199,16 +199,20 @@ please DO NOT file a public issue, instead send your report privately to
-with Kubernete's Team the upcoming security release timeline in such a way that Kubernetes 
+CoreDNS manage a list of "Collaborating Projects" that are assessed CoreDNS consumers and reliable secure projects. 
+That list aims to coordinate the security release in a controlled manner: 
+ 
```
In order to be transparent and for each project to be able to validate they are registered, should we create a public document CollaboratingProjects.md  with the public information regarding process and agreement with each project ?

fturib commented on README.md:
```diff
@@ -199,16 +199,20 @@ please DO NOT file a public issue, instead send your report privately to
-1. CoreDNS Team will provide backwards-compatible patch releases for all of the CoreDNS releases embedded into one of the 3 last releases of Kubernetes.
\ No newline at end of file
+   * ideally CoreDNS warns the Collaborating Projects one week before the security release date. 
+1. by default, CoreDNS Team expect to provide a security patch release on top of the current last release of CoreDNS. 
```
That last line is vague in purpose and let room for agreement on what versions will be patched - based on the needs defined during "coordination" of the security release.
NOTE: that is far from K8s expectation right now.

fturib commented:
NOTE: there is a specific document in [k8s about their security release process.](https://github.com/kubernetes/sig-release/blob/master/security-release-process-documentation/security-release-process.md). In that document there is a [specific paragraph on "private distributor list"](https://github.com/kubernetes/sig-release/blob/master/security-release-process-documentation/security-release-process.md#private-distributors-list) in order to coordinate the release.

Should we take that document as template, simplify it, and tune it for CoreDNS project ?

I also found this document on line, edited by CERT :  "[Guide to Coordinated Vulnerability Disclosure](https://resources.sei.cmu.edu/asset_files/SpecialReport/2017_003_001_503340.pdf)"

fturib commented:
I am working on it.
There is already a description on what is criteria to accept new distributor in the list - need some adaptation though.

miekg commented:
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] WIP - enhance..." ]
>I am working on it.
>There is already a description on what is criteria to accept new distributor in the list - need some adaptation though.

Ack, cool.

fturib commented on SECURITY-RELEASE-PROCESS.md:
```diff
@@ -0,0 +1,135 @@
+**Table of Contents**
+
+__TOC__
+
```
ok.. I expected a table of content here. Is there a way to automate it ?

fturib commented on SECURITY-RELEASE-PROCESS.md:
John is in PTO .. I could not ask him.

fturib commented:
I uploaded a new version based on existing security release process from k8s project.
I simplified as much as I can. Keeping all the part related to "private integrator list" and cooperation with projects using CoreDNS as a component,

Please feel free to amend, propose change, etc ..

fturib commented:
I had advises from our Security Department on how to improve this document. 
Will update soon, integrating advices.

fturib commented:
I update this document:
- includes now a definition of "Product Security Team" (need to define who is part of this Team in Owners.md) 
- a "Fix Team" build-up when a security release has to be made, with a "Team Lead" that have the responsibility to communicate and conduct the release.

NOTE: this new version is a merge of my former version (based on K8s document) and the same process defined in envoy project.

yongtang commented:
LGTM

johnbelamaric commented on SECURITY-RELEASE-PROCESS.md:
s/if/If/

fturib replied:
ok.

johnbelamaric commented on SECURITY-RELEASE-PROCESS.md:
I think you mean "compressed packet captures"?

fturib replied:
yes.

johnbelamaric commented on SECURITY-RELEASE-PROCESS.md:
s/an distributor/a distributor/

fturib replied:
ok

caniszczyk commented:
LGTM, let me know when you want CNCF to create the mailing list

fturib commented:
@johnbelamaric , @yongtang , @chrisohaver , @miekg : I added you in PST.
Please review and give your agreement (or not).

@miekg : who should create the needed mailing list on cncf.io (see TODO in the DESCRIPTION of this PR, and comment here : https://github.com/coredns/coredns/pull/1861#issuecomment-424373288)

fturib commented:
@miekg, @johnbelamaric : can you verify you are ok with the latest version ? Thanks.

johnbelamaric commented on SECURITY-RELEASE-PROCESS.md:
```diff
@@ -0,0 +1,188 @@
+# Security Release Process
+
+CoreDNS project has adopted this security disclosures and response policy
```
The CoreDNS project ... to ensure responsible handling of ...

fturib replied:
ok

johnbelamaric commented on SECURITY-RELEASE-PROCESS.md:
s/of set/of the set/

fturib replied:
done

johnbelamaric commented on SECURITY-RELEASE-PROCESS.md:
Rephrase this to:

If you find a security vulnerability or any security related issues,
please DO NOT file a public issue. Do not create a GitHub issue.
Instead, send your report privately to ...

Also, is that email address supposed to be the same as "product-security-team" or is it really a different address?

johnbelamaric commented on SECURITY-RELEASE-PROCESS.md:
s/If you know it, please/Please/

johnbelamaric commented on SECURITY-RELEASE-PROCESS.md:
/forget :/forget:/

johnbelamaric commented on SECURITY-RELEASE-PROCESS.md:
s/actively/active/

miekg commented on OWNERS:
```diff
@@ -22,13 +22,21 @@ approvers:
   - miekg
   - yongtang
 
+product-security-team:
```
this change can't be made to the OWNERS file just yet, because dreck doesn't support it

miekg replied:
I'll add some thing to dreck.

On second thought; does this need to be public?

fturib replied:
> this change can't be made to the OWNERS file just yet, because dreck doesn't support it
What is dreck ? You mean it needs to be ingested by a tool ?


> On second thought; does this need to be public?
You are right, I do not think it is mandatory to list the persons. But that would be more clear for everyone to know who is in this list and who is not.
NOTE: I just copied that behavior from other project.

miekg commented on README.md:
```diff
@@ -208,3 +208,5 @@ repository](https://github.com/coredns/deployment).
 issue, instead send your report privately to `security@coredns.io`. Security reports are greatly
 appreciated and we will publicly thank you for it.
+
+Please consult the page dedicated to [security vulnerability disclosures and security fix and release process](/SECURITY-RELEASE-PROCESS.md)
```
page?

fturib replied:
Right ... need to fix that.

miekg commented on SECURITY-RELEASE-PROCESS.md:
the @coredns.io is a bit adhoc at the moment and requires me manually adding people (or paying $$ for a actual group), this should probably be a closed group on groups.google.com or something hosted by the cnfc we people can subscribe.

miekg replied:
possibly aliased from distributors-announce@coredns.io

fturib replied:
Do you mean we need to ask creation of the dl for the 2 lists or only for distributor.
Something like:
- coredns-distributor-announce@googlegroup.com (to warn the distributors of early release)
- coredns-security-team@googlegroup.com (to discuss about security patch, only the PST team members)

and keep security@coredns.io for external request on 'security' ?

miekg commented on SECURITY-RELEASE-PROCESS.md:
Where is the hall of fame?

miekg commented on SECURITY-RELEASE-PROCESS.md:
extra whitespace

miekg commented:
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Security rele..." ]
>> this change can't be made to the OWNERS file just yet, because dreck doesn't support it
>What is dreck ? You mean it needs to be ingested by a tool ?

github.com/miekg/dreck the thing behind CorBot, yes this file is read by a tool.

>> On second thought; does this need to be public?
>You are right, I do not think it is mandatory to list the persons. But that would be more clear for everyone to know who is in this list and who is not.
>NOTE: I just copied that behavior from other project.

ack.

miekg commented:
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Security rele..." ]
>+* security@coredns.io : for any security concerns. Received by Product Security Team members, and used by this Team to discuss security issues and fixes.
>+* distributors-announce@coredns.io: for early private information on Security patch releases. see below how CoreDNS distributors can apply for this list.
>
>Do you mean we need to ask creation of the dl for the 2 lists or only for distributor.
>Something like:
>- coredns-distributor-announce@googlegroup.com (to warn the distributors of early release)
>- coredns-security-team@googlegroup.com (to discuss about security patch, only the PST team members)
>
>and keep security@coredns.io for external request on 'security' ?

No we can keep security@coredns.io as-is. We do need coredns-distributor-announce@googlegroup.com
and then make a forward from distributors-announce@coredns.io to point to that.
Does that work?

/Miek

--
Miek Gieben

fturib commented:
- I added a "security" entry in OWNER and corresponding PR for dreck : https://github.com/miekg/dreck/pull/107
- I removed unknown hall of fame
- the mailing list for announcing the early release will be : coredns-distributors-announce@googlegroup.com.

NOTE: when approval is ok. I will ask @caniszczyk if he can to create that private mailing list.

fturib commented:
@johnbelamaric : are you ok with this last version (if yes, please remove the "requested changes" flag)
@miekg : I updated after latests comments (and remove the list of security team in OWNERS). Let me know if ok. 
Thanks.

johnbelamaric commented:
/lgtm

miekg commented on OWNERS:
```diff
@@ -29,6 +29,7 @@ features:
   - branches
   - exec
 
+
```
this file shouldn't be touched by this PR anymore

fturib commented:
I updated. Just reverting the OWNER file that is no more involved in this PR.
